paddle/fluid/framework/reader.h

```diff
@@ -27,7 +27,7 @@ namespace framework {
 
 class ReaderBase {
  public:
-  virtual void ReadNext(std::vector<LoDTensor>* out);
+  virtual void ReadNext(std::vector<LoDTensor>* out, int dev_id = 0);
```

Comment:
I think maybe it is not a good idea to add `dev_id` to `ReaderBase::ReadNext`? A reader should not   care which device its data is inside.

Reply:
> I think maybe it is not a good idea to add dev_id to ReaderBase::ReadNext? A reader should not care which device its data is inside.

I will explain why..., and I haved tried remove `device_id` in reader, but maybe no better way:

Why there is and indepent queue for each device:
1. the double buffer would pop the item from the `LoDTensorBlockingQueue(queue)` and push to double buffer in ASYN. If there is only one queue, for the last batch, the double buffer from device0 may has two items, but device1 none, this would cause the training data unbalanced and nccl hung in parallel graph mode.
1.  Only one queue is the bottleneck for more than one devices.

And if we have multiple queues, the reader should know which to read. We bind the queue in startup_program by `create_py_reader_op`, but the startup_program is executed on single device instead of multiple deivces used in ParallelExecutor, so we can only know how much devices instead of which device to read,so maybe the  another way is execute the startup_program with ParallelExecutor, but it is also hard to compatible with the current implement.